File: prolog/old/tstamp.pl
:- writeln("-- tstamp.pl --"). 
:- ensure_loaded(["gb_env.pl","funcs.pl"]).
:- dynamic tsd/1.

tsd(D) :- D = _{},
   retractall(tsd(_)),
   assert(tsd(D)).

ts2(T, S, N2, N1) :-
    dstts(T, S, N2),
    srcts(S, N1).

prefn(Pre, N, Tsfn) :-
    pre(Pre, S1),
    string_concat(S1, N, Tsfn).

tsmakeexist(Fn) :-
    (   exists_file(Fn)
    ->  true
    ;   open(Fn, write, Str),
        close(Str)
    ).

savets :-
    pre('FLAGS', S1),
    string_concat(S1, "/tsd.pl", S2),
    open(S2, write, St1),
    tsd(D1),
    write_term(St1, tsd(D1), [fullstop(true), nl(true), quoted(true)]),
    close(St1).

tstime_m(Pre, N, Ft) :-
    tsd(D1),
    (   var(Ft)
    ->  catch(( ( '.'(D1, Pre, A),
                  '.'(A, N, B)
                ),
                Ft=B
              ),
              error(_, _),
              ( get_time(Ft),
                tstime_m(Pre, N, Ft)
              ))
    ;   retract(tsd(D1)),
        '.'(D1, put(Pre/N, Ft), C),
        D2=C,
        assert(tsd(D2)),
        savets
    ).

rtset :-
    findall(N1, (srcts(_, N1), rtset(N1)), _),
    savets.

rtset(N) :-
    get_time(Ct),
    tstime_m(rtbk, N, Ct).

rtset(N, Mt) :-
    tstime_m(rtbk, N, Mt).

bctck(N2, N1) :-
    tstime_m(rtbk, N1, Ft1),
    tstime_m(ct, N2, Ft2),
    Ft1>Ft2.

loadts :-
    pre('FLAGS', S1),
    string_concat(S1, "/tsd.pl", S2),
    (   exists_file(S2)
    ->  open(S2, read, St1),
        read_term(St1, tsd(D1), []),
        retractall(tsd(_)),
        assert(tsd(D1)),
        close(St1)
    ;   true
    ).

clr(N2, N1) :-
    tstime_m(rtbk, N1, Ft1),
    tstime_m(ct, N2, Ft2),
    (   Ft1=\=Ft2
    ->  tstime_m(ct, N2, Ft1),
        write("-clr"),
        writev(N2, 'N2'),
        writev(N1, 'N1'),
        dstts(T, S, N2),
        srcts(S, N1),
        writev(T, 'T'),
        writev(S, 'S'),
        nl
    ;   true
    ),
    nl.

bctclr(N2, N1) :-
    tstime_m(rtbk, N1, Ft1),
    tstime_m(ct, N2, Ft1),
    write("-bctclr"),
    writev(N2, 'N2'),
    writev(N1, 'N1'),
    nl.

:- loadts.

File: devbak/tstamp.py
import time
import pickle

from bkenv import *

tsd = None
tspf = pre['FLAGS'] / 'tsd.pp'


def tstime(pre, n, ft=None):
    global tsd
    if (pre, n) not in tsd:
        tsd[pre, n] = time.time()
        # print(pre, n, 'not found, injecting current time')
    oft = tsd[pre, n]
    if ft is not None:
        tsd[pre, n] = ft
    return oft


def n2ts(n2, n1):
    for (t, s) in dstts:
        if n2 == dstts[t, s]:
            if n1 == srcts[s]:
                return (t, s)


def clr(n2, n1):
    global tsd
    t1 = tstime('rtbk', n1)
    t2 = tstime('ct', n2)
    if t1 != t2:
        t, s = n2ts(n2, n1)
        print('-clr', n2, n1, t, s)
        tstime('ct', n2, t1)
        savets()
    return True


def loadts():
    global tsd
    try:
        tsd = {}
        with open(tspf, "rb") as fh:
            tsd1 = pickle.load(fh)
            for (pre, n) in tsd1:
                tstime(pre, n, tsd1[pre, n])
    except:
        tsd = {}


def savets():
    with open(tspf, "wb") as fh:
        pickle.dump(tsd, fh)


def bctck(N2, N1):
    Ft1 = tstime('rtbk', N1)
    Ft2 = tstime('ct', N2)
    return Ft1 > Ft2


def ts2(T, S):
    N2 = dstts[T, S]
    N1 = srcts[S]
    # print(T, S, N2, N1)
    return (N2, N1)


def rtset(N=None, Mt=None):
    if N is None:
        for (k, N1) in srcts.items():
            rtset(N1)
        return
    else:
        if Mt is None:
            tstime('rtbk', N, time.time())
            savets()
        else:
            tstime('rtbk', N, Mt)
            savets()


loadts()

if __name__ == '__main__':
    print(repr(tsd))
